src/shared/psi-util.c

```diff
@@ -0,0 +1,73 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <stdio.h>
+
+#include "alloc-util.h"
+#include "fileio.h"
+#include "psi-util.h"
+#include "stat-util.h"
+
+int read_resource_pressure(const char *path, PressureType type, ResourcePressure *ret) {
+        _cleanup_free_ char *value = NULL;
+        float avg10, avg60, avg300;
+        const char *field;
+        char *marker;
+        usec_t total;
+        char buf[5];
+        int r;
+
+        assert(path);
+        assert(ret);
+
+        if (type == SOME)
+                field = "some";
+        else if (type == FULL)
+                field = "full";
+
+        r = read_full_virtual_file(path, &value, NULL);
+        if (r < 0)
+                return r;
+
+        marker = value;
+
+        do {
+                r = sscanf(marker, "%5s avg10=%f avg60=%f avg300=%f total="USEC_FMT, buf, &avg10, &avg60, &avg300, &total);
```

Comment:
```
241/491 test-oomd-util                          FAIL     4.32 s (exit status 1)

--- command ---
21:03:35 SYSTEMD_KBD_MODEL_MAP='/build/src/locale/kbd-model-map' PATH='/build/build:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl' SYSTEMD_LANGUAGE_FALLBACK_MAP='/build/src/locale/language-fallback-map' /build/build/asan-wrapper-GSL.sh /build/build/test-oomd-util
--- stderr ---
=================================================================
==43050==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7f6b1e860395 at pc 0x7f6b23b157f3 bp 0x7ffeeab977c0 sp 0x7ffeeab96f70
WRITE of size 6 at 0x7f6b1e860395 thread T0
    #0 0x7f6b23b157f2  (/usr/lib/clang/9.0.1/lib/linux/libclang_rt.asan-x86_64.so+0x957f2)
    #1 0x7f6b23b167a7 in __isoc99_vsscanf (/usr/lib/clang/9.0.1/lib/linux/libclang_rt.asan-x86_64.so+0x967a7)
    #2 0x7f6b23b16897 in __isoc99_sscanf (/usr/lib/clang/9.0.1/lib/linux/libclang_rt.asan-x86_64.so+0x96897)
    #3 0x7f6b2329dbe3 in read_resource_pressure /build/build/../src/shared/psi-util.c:34:21
    #4 0x564db2956d1b in oomd_cgroup_context_get /build/build/../src/oom/oomd-util.c:277:13
    #5 0x564db2952f0b in test_oomd_cgroup_context_get /build/build/../src/oom/test-oomd-util.c:106:9
    #6 0x564db295260d in main /build/build/../src/oom/test-oomd-util.c:248:9
    #7 0x7f6b22b87022 in __libc_start_main (/usr/lib/libc.so.6+0x27022)
    #8 0x564db295252d in _start (/build/build/test-oomd-util+0x852d)

Address 0x7f6b1e860395 is located in stack of thread T0 at offset 149 in frame
    #0 0x7f6b2329da3f in read_resource_pressure /build/build/../src/shared/psi-util.c:10

  This frame has 6 object(s):
    [32, 40) 'value' (line 11)
    [64, 68) 'avg10' (line 12)
    [80, 84) 'avg60' (line 12)
    [96, 100) 'avg300' (line 12)
    [112, 120) 'total' (line 15)
    [144, 149) 'buf' (line 16) <== Memory access at offset 149 overflows this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow (/usr/lib/clang/9.0.1/lib/linux/libclang_rt.asan-x86_64.so+0x957f2) 
Shadow bytes around the buggy address:
  0x0fede3d04020: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x0fede3d04030: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x0fede3d04040: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x0fede3d04050: f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5 f5
  0x0fede3d04060: f1 f1 f1 f1 00 f2 f2 f2 04 f2 04 f2 04 f2 00 f2
=>0x0fede3d04070: f2 f2[05]f3 f3 f3 f3 f3 00 00 00 00 00 00 00 00
  0x0fede3d04080: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0fede3d04090: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0fede3d040a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0fede3d040b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0fede3d040c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==43050==ABORTING
```

Taken from https://ci.centos.org/job/systemd-pr-build-vagrant-sanitizers/4479/artifact//systemd-centos-ci/artifacts_all/artifacts_1Gq7Ke/vagrant-logs.FSn/vagrant-arch-sanitizers-clang-testsuite.e51/ninja-test_sanitizers_FAIL.log